### PR TITLE
add autoscaling_group

### PR DESCRIPTION
**Pull request by mnkap**

### CODE REVIEW

File: level1/network.tf [hunk 1]
```diff
@@ -7,8 +7,14 @@ resource "aws_vpc" "main" {
...
 
+data "aws_availability_zones" "available" {
+  state = "available"
+
+}
+
 resource "aws_subnet" "public" {
-  count = length(var.public_cidr)
+  availability_zone = data.aws_availability_zones.available.names[0]
+  count             = length(var.public_cidr)
```
Comment (vladimirmukhin) [inline level1/network.tf:17]: count should always be the first argument and also should be separated by a newline from other arguments

File: level1/network.tf [hunk 2]
```diff
@@ -19,7 +25,8 @@ resource "aws_subnet" "public" {
 }
 
 resource "aws_subnet" "private" {
-  count = length(var.private_cidr)
+  availability_zone = data.aws_availability_zones.available.names[1]
+  count             = length(var.private_cidr)
```
Comment (vladimirmukhin) [inline level1/network.tf:29]: count should always be the first argument and also should be separated by a newline from other arguments

File: level1/network.tf [hunk 3]
```diff
@@ -19,7 +25,8 @@ resource "aws_subnet" "public" {
 }
 
 resource "aws_subnet" "private" {
-  count = length(var.private_cidr)
+  availability_zone = data.aws_availability_zones.available.names[1]
```
Comment (vladimirmukhin) [inline level1/network.tf:28]:
```suggestion
  availability_zone = data.aws_availability_zones.available.names[count.index]
```

File: level1/network.tf [hunk 4]
```diff
@@ -7,8 +7,14 @@ resource "aws_vpc" "main" {
...
 }
 
+data "aws_availability_zones" "available" {
+  state = "available"
+
+}
+
 resource "aws_subnet" "public" {
-  count = length(var.public_cidr)
+  availability_zone = data.aws_availability_zones.available.names[0]
```
Comment (vladimirmukhin) [inline level1/network.tf:16]:
```suggestion
  availability_zone = data.aws_availability_zones.available.names[count.index]
```

File: level1/network.tf [hunk 5]
```diff
@@ -7,8 +7,14 @@ resource "aws_vpc" "main" {
   }
 }
 
+data "aws_availability_zones" "available" {
+  state = "available"
+
+}
```
Comment (vladimirmukhin) [inline level1/network.tf:13]:
```suggestion
  state = "available"
}
```

File: level2/alb.tf [hunk 6]
```diff
@@ -30,6 +30,7 @@ resource "aws_lb" "main" {
   security_groups    = [aws_security_group.load_balancer.id]
   subnets            = data.terraform_remote_state.level1.outputs.public_subnet_id
 
+
```
Comment (vladimirmukhin) [inline level2/alb.tf:33]: Why did you add this newline?

File: level2/asg.tf [hunk 7]
```diff
@@ -0,0 +1,25 @@
+resource "aws_launch_configuration" "main" {
+  name_prefix     = "${var.env_code}-"
+  image_id        = data.aws_ami.amazonlinux.id
+  instance_type   = "t2.micro"
+  security_groups = [aws_security_group.private.id]
+  user_data       = file("user-data.sh")
+  key_name        = "Main"
```
Comment (vladimirmukhin) [inline level2/asg.tf:7]:
```suggestion
  key_name        = "main"
```

File: level2/asg.tf [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+resource "aws_autoscaling_group" "main" {
+  name             = var.env_code
+  desired_capacity = 2
+  max_size         = 2
+  min_size         = 2
+
+  target_group_arns    = [aws_lb_target_group.main.arn]
+  launch_configuration = aws_launch_configuration.main.name
+  vpc_zone_identifier  = data.terraform_remote_state.level1.outputs.private_subnet_id
```
Comment (vladimirmukhin) [inline level2/asg.tf:18]: It will create all your instances in a single availability zone. You need instances to span multiple availability zones

**Review by vladimirmukhin — CHANGES_REQUESTED** [review]

File: level2/asg.tf (hunk 7 above)
Comment (mnkap) [reply level2/asg.tf:7]: The key name is valid

File: level2/asg.tf (hunk 8 above)
Comment (mnkap) [reply level2/asg.tf:18]: `vpc_zone_identifier  = [data.terraform_remote_state.level1.outputs.private_subnet_id, data.terraform_remote_state.level1.outputs.public_subnet_id]`

File: level2/asg.tf (hunk 7 above)
Comment (vladimirmukhin) [reply level2/asg.tf:7]: Please use lowercase

**Review by vladimirmukhin — APPROVED** [review]

**Pull request merged**